x-pack/plugins/security_solution/public/common/components/navigation/solution_grouped_nav/solution_grouped_nav.tsx

```diff
@@ -0,0 +1,167 @@
+
+  const [activePortalNavId, setActivePortalNavId] = useState<ActivePortalNav>(null);
+  const activePortalNavIdRef = useRef<ActivePortalNav>(null);
+
+  const openPortalNav = (navId: string) => {
+    activePortalNavIdRef.current = navId;
+    setActivePortalNavId(navId);
+  };
+
+  const closePortalNav = () => {
+    activePortalNavIdRef.current = null;
+    setActivePortalNavId(null);
+  };
+
+  const onClosePortalNav = useCallback(() => {
+    const currentPortalNavId = activePortalNavIdRef.current;
+    setTimeout(() => {
+      // This event is triggered on outside click.
+      // Closing the side nav at the end of event loop to make sure it
+      // closes also if the active "nav group" button has been clicked (toggle),
+      // but it does not close if any some other "nav group" open button has been clicked.
+      if (activePortalNavIdRef.current === currentPortalNavId) {
+        closePortalNav();
+      }
+    });
+  }, []);
+
+  const navItemsById = useMemo(
+    () =>
+      [...items, ...footerItems].reduce<
+        Record<string, { title: string; subItems: PortalNavItem[] }>
+      >((acc, navItem) => {
+        if (isDefaultNavItem(navItem) && navItem.items && navItem.items.length > 0) {
+          acc[navItem.id] = {
+            title: navItem.label,
+            subItems: navItem.items,
+          };
+        }
+        return acc;
+      }, {}),
+    [items, footerItems]
+  );
+
+  const portalNav = useMemo(() => {
+    if (activePortalNavId == null || !navItemsById[activePortalNavId]) {
+      return null;
+    }
+    const { subItems, title } = navItemsById[activePortalNavId];
+    return <SolutionGroupedNavPanel onClose={onClosePortalNav} items={subItems} title={title} />;
+  }, [activePortalNavId, navItemsById, onClosePortalNav]);
+
+  const renderNavItem = useCallback(
+    (navItem: NavItem) => {
+      if (isCustomNavItem(navItem)) {
+        return <Fragment key={navItem.id}>{navItem.render()}</Fragment>;
+      }
+      const { id, href, label, onClick } = navItem;
+      const isActive = activePortalNavId === id;
+      const isCurrentNav = selectedId === id;
+
+      const itemClassNames = classNames('solutionGroupedNavItem', {
+        'solutionGroupedNavItem--isActive': isActive,
+        'solutionGroupedNavItem--isPrimary': isCurrentNav,
+      });
+      const buttonClassNames = classNames('solutionGroupedNavItemButton');
+
+      return (
+        // eslint-disable-next-line @elastic/eui/href-or-on-click
+        <EuiLink
+          key={id}
+          href={href}
+          onClick={onClick}
+          color={isCurrentNav ? 'primary' : 'text'}
+          data-test-subj={`groupedNavItemLink-${id}`}
+        >
+          <EuiListGroupItemStyled
+            className={itemClassNames}
+            isActive={isActive}
+            color={isCurrentNav ? 'primary' : 'text'}
+            label={label}
+            size="s"
+            {...(!isMobileSize && navItemsById[id] != null
```

Comment:
worth moving this out into a useMemo?

Reply:
we are inside a `useCallback` here, which is called in a map for each link item, don't think it would be easy to extract this part.